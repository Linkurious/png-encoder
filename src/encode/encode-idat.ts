/**
 * @since 2019-12-30 01:47
 * @author vivaxy
 */
import * as pako from 'pako';
import rescaleSample from '../helpers/rescale-sample';
import { concatUInt8Array } from '../helpers/typed-array';
import { channelToTypedArray } from '../helpers/channels';
import { filters, FILTER_TYPES } from '../helpers/filters';
import {
  COLOR_TYPES,
  COLOR_TYPES_TO_CHANNEL_PER_PIXEL,
} from '../helpers/color-types';

export default function encodeIDAT(
  data: number[],
  width: number,
  colorType: COLOR_TYPES,
  depth: number,
  interlace: number,
  palette?: [number, number, number, number][],
) {
  let typedArray = new Uint8Array();
  const scanlineCount = (data.length / width) >> 2;
  const channelPerPixel = COLOR_TYPES_TO_CHANNEL_PER_PIXEL[colorType];
  const scanlineWidth = (width * channelPerPixel * depth + 7) >> 3;
  const bitPerPixel = channelPerPixel * depth;
  const bytePerPixel = bitPerPixel >> 3;
  let prevUnfilteredLine = new Uint8Array();

  for (let scanlineIndex = 0; scanlineIndex < scanlineCount; scanlineIndex++) {
    // each line
    const channels: number[] = [];
    const scanlineStartIndex = (scanlineIndex * width) << 2;
    for (let pixelIndex = 0; pixelIndex < width; pixelIndex++) {
      // each pixel
      const pixelStartIndex = scanlineStartIndex + (pixelIndex << 2);
      const pixel = data.slice(pixelStartIndex, pixelStartIndex + 4);

      // channels
      if (colorType === COLOR_TYPES.GRAYSCALE) {
        channels.push(rescaleSample(pixel[0], 8, depth));
      } else if (colorType === COLOR_TYPES.TRUE_COLOR) {
        channels.push(
          rescaleSample(pixel[0], 8, depth),
          rescaleSample(pixel[1], 8, depth),
          rescaleSample(pixel[2], 8, depth),
        );
      } else if (colorType === COLOR_TYPES.PALETTE) {
        if (!palette) {
          throw new Error('Palette is required');
        }
        let paletteIndex = -1;
        for (let i = 0; i < palette.length; i++) {
          const paletteColor = palette[i];
          if (
            paletteColor[0] === pixel[0] &&
            paletteColor[1] === pixel[1] &&
            paletteColor[2] == pixel[2] &&
            paletteColor[3] === pixel[3]
          ) {
            paletteIndex = i;
          }
        }
        if (paletteIndex === -1) {
          throw new Error(
            'Invalid palette. Color (' +
              pixel.join(', ') +
              ') is not in palette',
          );
        }
        channels.push(paletteIndex);
      } else if (colorType === COLOR_TYPES.GRAYSCALE_WITH_ALPHA) {
        channels.push(
          rescaleSample(pixel[0], 8, depth),
          rescaleSample(pixel[3], 8, depth),
        );
      } else if (colorType === COLOR_TYPES.TRUE_COLOR_WITH_ALPHA) {
        channels.push(
          rescaleSample(pixel[0], 8, depth),
          rescaleSample(pixel[1], 8, depth),
          rescaleSample(pixel[2], 8, depth),
          rescaleSample(pixel[3], 8, depth),
        );
      }

      // set channels to unfilteredLine
    }
    const unfilteredLine = channelToTypedArray(channels, depth, scanlineWidth);
    // filter line
    let minFilterSum = Infinity;
    let filteredLine = new Uint8Array();
    let filterType = 0;
    const filterTypes = [
      FILTER_TYPES.NONE,
      FILTER_TYPES.SUB,
      FILTER_TYPES.UP,
      FILTER_TYPES.AVERAGE,
      FILTER_TYPES.PAETH,
    ];
    for (let filterIndex = 0; filterIndex < filterTypes.length; filterIndex++) {
      const filter = filters[filterTypes[filterIndex] as FILTER_TYPES];
      const { sum, filtered } = filter(
        unfilteredLine,
        bytePerPixel,
        prevUnfilteredLine,
      );
      if (sum < minFilterSum) {
        minFilterSum = sum;
        filteredLine = filtered;
        filterType = filterTypes[filterIndex];
      }
    }

    prevUnfilteredLine = unfilteredLine;

    typedArray = concatUInt8Array(
      typedArray,
      concatUInt8Array(new Uint8Array([filterType]), filteredLine),
    );
  }
  return pako.deflate(typedArray);
}
