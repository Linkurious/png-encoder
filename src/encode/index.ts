/**
 * @since 2019-10-30 03:01
 * @author vivaxy
 */
import crc32 from '../helpers/crc32';
import Metadata from '../helpers/metadata';
import PNG_SIGNATURE from '../helpers/signature';
import { concatUInt8Array } from '../helpers/typed-array';

export default function encode(metadata: Metadata) {
  // Signature
  let typedArray = new Uint8Array(PNG_SIGNATURE);

  // Helpers
  function packUInt32BE(value: number) {
    return new Uint8Array([
      (value >>> 24) & 0xff,
      (value >>> 16) & 0xff,
      (value >>> 8) & 0xff,
      value & 0xff,
    ]);
  }

  function packUInt8(value: number) {
    return new Uint8Array([value & 0xff]);
  }

  function packChunkName(name: string) {
    const data = new Uint8Array(name.length);
    for (let i = 0; i < name.length; i++) {
      data[i] = name.charCodeAt(i);
    }
    return data;
  }

  // Chunks
  const chunkPackers: {
    [chunkName: string]: () => Uint8Array;
  } = {
    IHDR: packIHDR,
    PLTE: packPLTE,
    IDAT: packIDAT,
    IEND: packIEND,
  };

  function packIHDR() {
    let data = new Uint8Array();
    data = concatUInt8Array(data, packUInt32BE(metadata.width));
    data = concatUInt8Array(data, packUInt32BE(metadata.height));
    data = concatUInt8Array(data, packUInt32BE(metadata.height));
    data = concatUInt8Array(data, packUInt8(metadata.depth));
    data = concatUInt8Array(data, packUInt8(metadata.colorType));
    data = concatUInt8Array(data, packUInt8(metadata.compression));
    data = concatUInt8Array(data, packUInt8(metadata.filter));
    data = concatUInt8Array(data, packUInt8(metadata.interlace));
    return data;
  }

  function packPLTE() {
    let data = new Uint8Array();
    return data;
  }

  function packIDAT() {
    let data = new Uint8Array();
    return data;
  }

  function packIEND() {
    let data = new Uint8Array();
    return data;
  }

  Object.keys(chunkPackers).forEach(function(chunkName) {
    const nameData = packChunkName(chunkName);
    const data = chunkPackers[chunkName]();
    const lengthData = packUInt32BE(data.length);
    const lengthAndData = concatUInt8Array(lengthData, data);
    const calculatedCrc32 = crc32(lengthAndData);
    const endData = packUInt32BE(calculatedCrc32);

    const chunkData = concatUInt8Array(
      concatUInt8Array(nameData, lengthAndData),
      endData,
    );
    typedArray = concatUInt8Array(typedArray, chunkData);
  });

  return typedArray;
}
