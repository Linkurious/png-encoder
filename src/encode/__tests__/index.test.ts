/**
 * @since 20180911 17:16
 * @author vivaxy
 * @input input.json
 * @todo filter-1
 * @todo filter-2
 * @todo filter-3
 * @todo filter-4
 * @todo bit-depth-04
 * @todo bit-depth-08
 * @todo bit-depth-16
 * @todo chunk-bKGD
 * @todo chunk-cHRM
 * @todo chunk-gMMA
 * @todo chunk-hIST
 * @todo chunk-IDAT-multiple
 * @todo chunk-IDAT-single
 * @todo chunk-IEND
 * @todo chunk-IHDR
 * @todo chunk-iTXt
 * @todo chunk-pHYs
 * @todo chunk-PLTE
 * @todo chunk-sBIT
 * @todo chunk-sPLT
 * @todo chunk-sRGB
 * @todo chunk-tEXt
 * @todo chunk-tIME
 * @todo chunk-tRNS-color-type-0
 * @todo chunk-tRNS-color-type-2
 * @todo chunk-tRNS-color-type-3
 * @todo chunk-zTXt
 * @todo color-type-0
 * @todo color-type-2
 * @todo color-type-3
 * @todo color-type-4
 * @todo color-type-6
 * @todo interlace-8x-size
 * @todo interlace-odd-size
 */
import encode from '..';

// @ts-ignore
test.each(global.testcases)('decode %s', async function(
  testcaseName,
  metadataPath,
) {
  const metadata = require(metadataPath);
  expect(encode(metadata)).toMatchSnapshot();
});
