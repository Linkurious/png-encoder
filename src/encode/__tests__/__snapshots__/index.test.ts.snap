// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decode bit-depth-01 1`] = `
Uint8Array [
  137,
  80,
  78,
  71,
  13,
  10,
  26,
  10,
  0,
  0,
  0,
  13,
  73,
  72,
  68,
  82,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  1,
  1,
  3,
  0,
  0,
  0,
  37,
  219,
  86,
  202,
  0,
  0,
  0,
  1,
  115,
  82,
  71,
  66,
  1,
  217,
  201,
  44,
  127,
  0,
  0,
  0,
  9,
  112,
  72,
  89,
  115,
  0,
  0,
  11,
  19,
  0,
  0,
  11,
  19,
  1,
  0,
  154,
  156,
  24,
  0,
  0,
  0,
  3,
  80,
  76,
  84,
  69,
  0,
  0,
  0,
  167,
  122,
  61,
  218,
  0,
  0,
  0,
  10,
  73,
  68,
  65,
  84,
  120,
  156,
  99,
  96,
  0,
  0,
  0,
  2,
  0,
  1,
  72,
  175,
  164,
  113,
  0,
  0,
  0,
  0,
  73,
  69,
  78,
  68,
  174,
  66,
  96,
  130,
]
`;

exports[`decode bit-depth-02 1`] = `
Uint8Array [
  137,
  80,
  78,
  71,
  13,
  10,
  26,
  10,
  0,
  0,
  0,
  13,
  73,
  72,
  68,
  82,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  2,
  2,
  3,
  0,
  0,
  0,
  15,
  216,
  229,
  183,
  0,
  0,
  0,
  1,
  115,
  82,
  71,
  66,
  1,
  217,
  201,
  44,
  127,
  0,
  0,
  0,
  9,
  112,
  72,
  89,
  115,
  0,
  0,
  11,
  19,
  0,
  0,
  11,
  19,
  1,
  0,
  154,
  156,
  24,
  0,
  0,
  0,
  12,
  80,
  76,
  84,
  69,
  255,
  0,
  0,
  0,
  255,
  0,
  0,
  0,
  255,
  255,
  255,
  255,
  251,
  0,
  96,
  246,
  0,
  0,
  0,
  4,
  116,
  82,
  78,
  83,
  255,
  255,
  255,
  127,
  128,
  144,
  197,
  89,
  0,
  0,
  0,
  12,
  73,
  68,
  65,
  84,
  120,
  156,
  99,
  16,
  96,
  90,
  0,
  0,
  0,
  216,
  0,
  179,
  171,
  91,
  31,
  104,
  0,
  0,
  0,
  0,
  73,
  69,
  78,
  68,
  174,
  66,
  96,
  130,
]
`;

exports[`decode filter-0 1`] = `
Uint8Array [
  137,
  80,
  78,
  71,
  13,
  10,
  26,
  10,
  0,
  0,
  0,
  13,
  73,
  72,
  68,
  82,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  2,
  2,
  3,
  0,
  0,
  0,
  15,
  216,
  229,
  183,
  0,
  0,
  0,
  1,
  115,
  82,
  71,
  66,
  1,
  217,
  201,
  44,
  127,
  0,
  0,
  0,
  9,
  112,
  72,
  89,
  115,
  0,
  0,
  11,
  19,
  0,
  0,
  11,
  19,
  1,
  0,
  154,
  156,
  24,
  0,
  0,
  0,
  12,
  80,
  76,
  84,
  69,
  255,
  0,
  0,
  0,
  255,
  0,
  0,
  0,
  255,
  255,
  255,
  255,
  251,
  0,
  96,
  246,
  0,
  0,
  0,
  4,
  116,
  82,
  78,
  83,
  255,
  255,
  255,
  127,
  128,
  144,
  197,
  89,
  0,
  0,
  0,
  12,
  73,
  68,
  65,
  84,
  120,
  156,
  99,
  16,
  96,
  90,
  0,
  0,
  0,
  216,
  0,
  179,
  171,
  91,
  31,
  104,
  0,
  0,
  0,
  0,
  73,
  69,
  78,
  68,
  174,
  66,
  96,
  130,
]
`;
